### PR TITLE
feat: Add sorted set put and fetch functions

### PR DESCRIPTION
**Pull request by nand4011**
Add sorted set put element(s), fetch by rank, and fetch by score.

Add create and delete cache methods for testing.

Add test setup and teardown scripts so that we can guarantee the cache exists for doc and integration tests, and ensure it is deleted after they run.

Call the setup and teardown scripts in github ci.

### CODE REVIEW

File: src/requests/cache/create_cache.rs
Comment (nand4011) [inline src/requests/cache/create_cache.rs:23]: Having to call to_string or when giving strings to these methods feels strange. I tried changing them to &str, but then I need to define a lifetime for the request struct. What is the idiomatic thing to do here?

File: scripts/test-setup.rs
Comment (cprice404) [inline scripts/test-setup.rs:10]: any thoughts on providing a default here? it kinda always bugs me for local dev when we have to add another env var, but i can imagine some pros and cons

File: scripts/test-setup.rs
Comment (cprice404) [inline scripts/test-setup.rs:13]: supernit: i've been slowly trying to standardize on MOMENTO_API_KEY for these. In the future I can imagine that we might have a default credentialprovider that looks for creds in a few well known places, including a well known env var name, like AWS does. I don't see any super compelling reason to make tests use a different env var name. Thoughts?

File: src/requests/cache/create_cache.rs
Comment (cprice404) [reply src/requests/cache/create_cache.rs:23]: I don't think `to_string` is uncommon but I think there might also be a way to specify a signature that can accept both, using some kind of `Into` thing. I'll create a ticket to look into it but we don't need to block this pr on it.

File: src/requests/cache/set_add_elements.rs
Comment (cprice404) [inline src/requests/cache/set_add_elements.rs:47]: thanks for adding this.

do you have any strong opinions on how we name these going forward? I was waffling between using the `with_` prefix or not. I might be leaning towards not using that prefix, because in some stuff I've worked with in the past, it was used to signify a copy constructor (thus, leaving the original object intact without mutating it, whereas this is definitely mutating it).

File: src/requests/cache/sorted_set_fetch_by_rank.rs
Comment (cprice404) [inline src/requests/cache/sorted_set_fetch_by_rank.rs:43]: do you know if this will get excluded from the generated docs, or whether the comment blocks have to be contiguous at the top/bottom of the function in order to get excluded? if you don't know we can just look into this as a follow-up, not a blocker for this PR.

File: src/requests/cache/sorted_set_fetch_by_rank.rs
Comment (cprice404) [inline src/requests/cache/sorted_set_fetch_by_rank.rs:55]: this looks pretty good to me! same question about `with_` prefix but otherwise I think I like how it looks

File: src/requests/cache/sorted_set_put_element.rs
Comment (cprice404) [inline src/requests/cache/sorted_set_put_element.rs:46]: nit I think the convention for naming these generic types is to name them the first letter of whatever field they are going into, so we did `E` for "elements", and this one would be `V` I think.

File: src/requests/cache/sorted_set_put_elements.rs
Comment (cprice404) [inline src/requests/cache/sorted_set_put_elements.rs:47]: I can imagine that we might want to support a Map type as input here, so we might need another Into thing, but we can explore that in a follow-up PR.

File: test-util/src/lib.rs
Comment (cprice404) [inline test-util/src/lib.rs:59]: same comments here about whether it would be sane to use a default for test_cache_name if it's missing, and whether we should just use MOMENTO_API_KEY

File: tests/cache_sorted_set_tests.rs
Comment (cprice404) [inline tests/cache_sorted_set_tests.rs:13]: so, one thing I wanted to sync with you on, it seems like in a lot of rust code the tests are just added to the same file as the actual code that is being tested. idk if we have prior art on that in this repo or if that is sane, or if for us it makes more sense to keep them separate especially since they are integration tests. any thoughts?

File: tests/cache_test_state.rs
Comment (cprice404) [inline tests/cache_test_state.rs:69]: wow, fancy. I understand how this is working but am a little surprised that we have to use a channel to pass it back? naively, it seems like with an Arc we should be able to just reference the same client from multiple threads, but maybe Arcs are just for referencing from multiple functions in the same thread, and channels are required to move across thread boundaries?

anyway, I assume you and kenny paired on this and aligned on this solution, in which case I'm +1, just musing about how it reads to a newb :)

File: tests/cache_test_state.rs
Comment (cprice404) [inline tests/cache_test_state.rs:5]: did you guys talk about the different crates that are available for this, and just curious if there was a reason this one seemed best?

**Review by cprice404 — COMMENTED** [review]
this looks great. left a bunch of questions that i'm interested in your feedback on, but no blockers, and i'm happy to pick up any changes that we agree should go into this PR before it gets merged so that you can stay focused on other tasks.

File: .github/workflows/ci.yaml
Comment (nand4011) [reply .github/workflows/ci.yaml:54]: The setup and teardown binaries are defined as examples in Cargo.toml. It is hacky to do that, but normal bins don't have access to dev dependencies, so it would be hard to get access to tokio without locking that dependency behind a feature flag. There might be a better way to organize that, but I couldn't find one.

File: .github/workflows/ci.yaml
Comment (nand4011) [reply .github/workflows/ci.yaml:62]: It will always run even if the job fails or is cancelled. I wanted to make sure that we wouldn't leak caches in those cases. I haven't tested it properly since the tests have always passed, however.

File: scripts/test-setup.rs
Comment (nand4011) [reply scripts/test-setup.rs:10]: That also reminds me that I forgot to make a CONTRIBUTING.md.

File: scripts/test-setup.rs
Comment (nand4011) [reply scripts/test-setup.rs:13]: I'm good with MOMENTO_API_KEY. I agree that there isn't much reason to separate them, other than if someone consistently needed to run a local application using a different key than they were using to develop an sdk.

File: src/requests/cache/set_add_elements.rs
Comment (nand4011) [reply src/requests/cache/set_add_elements.rs:47]: I'm not too attached to it, but I would need to think of another name that felt good being called in a chain, like
```
.with_order(SortOrder::Ascending)
.with_start_rank(1)
.with_end_rank(3);
```
I suppose we could just do
```
.order(SortOrder::Ascending)
.start_rank(1)
.end_rank(3);
```
I also think it would be worth seeing if this style is more ergonomic for building config objects than the explicit builders we're currently using.

File: src/requests/cache/sorted_set_fetch_by_rank.rs
Comment (nand4011) [reply src/requests/cache/sorted_set_fetch_by_rank.rs:43]: It will exclude any line maked with a hash. Here is what that example looks like:
```
use momento::requests::cache::sorted_set_fetch_by_rank::SortOrder;
use momento::requests::cache::sorted_set_fetch_by_rank::SortedSetFetchByRankRequest;
use momento::response::cache::sorted_set_fetch::SortedSetFetch;
let sorted_set_name = "sorted_set";

let put_element_response = cache_client.sorted_set_put_elements(
    cache_name.to_string(),
    sorted_set_name.to_string(),
    vec![("value1", 1.0), ("value2", 2.0), ("value3", 3.0), ("value4", 4.0)]
).await?;
```

File: src/requests/cache/sorted_set_put_element.rs
Comment (nand4011) [reply src/requests/cache/sorted_set_put_element.rs:46]: Good find. That's a copy/paste error from when I copied it from set_add_elements.

File: tests/cache_sorted_set_tests.rs
Comment (nand4011) [reply tests/cache_sorted_set_tests.rs:13]: I honestly hadn't thought about that. I just saw we had a tests directory with some integration tests in it and my Java instincts kicked in. I would need to do an experiment to see if the global test state would still work the same way, but I'm good with adding them to the same file. My only concern would be if it would become hard to read if the cache client were filled with doctests and integration tests.

File: tests/cache_test_state.rs
Comment (nand4011) [reply tests/cache_test_state.rs:69]: Yeah, Kenny and I went through several frustrating steps to get to this point. There may be an easier solution, but I don't currently have the Rust chops to point it out.

File: .github/workflows/ci.yaml
Comment (cprice404) [reply .github/workflows/ci.yaml:54]: nope that seems fine but we might want to put a comment here. i can do it.

File: scripts/test-setup.rs
Comment (cprice404) [reply scripts/test-setup.rs:10]: discussed in slack, the current behavior seems okay, will standardize on a default fallback cache name

File: scripts/test-setup.rs
Comment (cprice404) [reply scripts/test-setup.rs:13]: thx

File: tests/cache_sorted_set_tests.rs
Comment (cprice404) [reply tests/cache_sorted_set_tests.rs:13]: k, will fish around for opinions on the backend team... on this one i could make a case either way.

File: src/requests/cache/set_add_elements.rs
Comment (cprice404) [reply src/requests/cache/set_add_elements.rs:47]: yeah i was kind of thinking of just removing the `with_`, just like you did in your example there.

re: config builders, yeah i will have to dig into that more to form an opinion... in my head it boils down to whether the struct for the builder is the same struct for the final object or not. will keep it in mind though.

File: src/requests/cache/sorted_set_fetch_by_rank.rs
Comment (cprice404) [reply src/requests/cache/sorted_set_fetch_by_rank.rs:43]: sweet, thanks

File: tests/cache_test_state.rs
Comment (cprice404) [reply tests/cache_test_state.rs:5]: if it's in std then why do we need the dep on once_cell? is the std thing only in nightly or something?

File: tests/cache_test_state.rs
Comment (cprice404) [reply tests/cache_test_state.rs:69]: ack

File: src/requests/cache/set_add_elements.rs
Comment (cprice404) [reply src/requests/cache/set_add_elements.rs:47]: will create a separate ticket for this

File: tests/cache_test_state.rs
Comment (cprice404) [reply tests/cache_test_state.rs:5]: discussed in slack - confirmed that the thing we need here is only in nightly, so we still need the crate

**Comment by cprice404** [comment]
Chris TODO:
- [ ] add comments re: `cargo --example`
- [ ] add comments re: `always()`
- [ ] add default cache name fallback and maybe DRY up?
- [ ] change TEST_API_KEY to MOMENTO_API_KEY and maybe DRY up?
- [ ] change relevant `E` generics to `V`s

**Comment by cprice404** [comment]
@nand4011 I've touched up everything that I was interested in for this PR, and made some tickets for followup items. Would you mind giving it a peek?

**Review by cprice404 — APPROVED** [review]

**Comment by nand4011** [comment]
I can't approve it, but it LGTM.

**Pull request merged**
